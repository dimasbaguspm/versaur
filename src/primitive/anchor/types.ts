/**
 * AnchorProps defines the props for the Anchor component
 * @see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a
 */
import type { AnchorHTMLAttributes, ReactNode } from 'react'

export interface AnchorProps extends AnchorHTMLAttributes<HTMLAnchorElement> {
  /**
   * The content to be rendered inside the anchor
   */
  children: ReactNode
  /**
   * Color role for the anchor (default: 'primary')
   */
  color?: 'primary' | 'secondary' | 'ghost' | 'danger' | 'neutral'
  /**
   * Font size utility (default: 'base')
   * See: https://tailwindcss.com/docs/font-size
   */
  fontSize?:
    | 'xs'
    | 'sm'
    | 'base'
    | 'lg'
    | 'xl'
    | '2xl'
    | '3xl'
    | '4xl'
    | '5xl'
    | '6xl'
    | '7xl'
    | '8xl'
    | '9xl'
    | 'inherit'
    | undefined
  /**
   * Font weight utility (default: 'medium')
   * See: https://tailwindcss.com/docs/font-weight
   */
  fontWeight?:
    | 'thin'
    | 'extralight'
    | 'light'
    | 'normal'
    | 'medium'
    | 'semibold'
    | 'bold'
    | 'extrabold'
    | 'black'
    | 'inherit'
    | undefined
  /**
   * If true, removes underline style from anchor
   */
  quiet?: boolean
}
