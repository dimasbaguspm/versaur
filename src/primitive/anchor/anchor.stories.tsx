/**
 * Anchor stories for Storybook
 * Demonstrates all color roles, fontSize, fontWeight
 */
import type { Meta, StoryObj } from '@storybook/react'
import { Anchor } from './anchor'

const meta: Meta<typeof Anchor> = {
  title: 'Primitive/Anchor',
  component: Anchor,
  tags: ['autodocs'],
  parameters: {
    layout: 'centered',
  },
}
export default meta

type Story = StoryObj<typeof Anchor>

export const Primary: Story = {
  args: {
    href: '#',
    children: 'Primary Anchor',
    color: 'primary',
    fontSize: 'base',
    fontWeight: 'medium',
    quiet: false,
  },
}

export const PrimaryQuiet: Story = {
  args: {
    href: '#',
    children: 'Primary Quiet Anchor',
    color: 'primary',
    fontSize: 'base',
    fontWeight: 'medium',
    quiet: true,
  },
}

export const Secondary: Story = {
  args: {
    href: '#',
    children: 'Secondary Anchor',
    color: 'secondary',
    fontSize: 'lg',
    fontWeight: 'bold',
    quiet: false,
  },
}

export const SecondaryQuiet: Story = {
  args: {
    href: '#',
    children: 'Secondary Quiet Anchor',
    color: 'secondary',
    fontSize: 'lg',
    fontWeight: 'bold',
    quiet: true,
  },
}

export const Ghost: Story = {
  args: {
    href: '#',
    children: 'Ghost Anchor',
    color: 'ghost',
    fontSize: 'sm',
    fontWeight: 'normal',
    quiet: false,
  },
}

export const GhostQuiet: Story = {
  args: {
    href: '#',
    children: 'Ghost Quiet Anchor',
    color: 'ghost',
    fontSize: 'sm',
    fontWeight: 'normal',
    quiet: true,
  },
}

export const Danger: Story = {
  args: {
    href: '#',
    children: 'Danger Anchor',
    color: 'danger',
    fontSize: 'xl',
    fontWeight: 'bold',
    quiet: false,
  },
}

export const DangerQuiet: Story = {
  args: {
    href: '#',
    children: 'Danger Quiet Anchor',
    color: 'danger',
    fontSize: 'xl',
    fontWeight: 'bold',
    quiet: true,
  },
}

export const Neutral: Story = {
  args: {
    href: '#',
    children: 'Neutral Anchor',
    color: 'neutral',
    fontSize: 'base',
    fontWeight: 'normal',
    quiet: false,
  },
}

export const NeutralQuiet: Story = {
  args: {
    href: '#',
    children: 'Neutral Quiet Anchor',
    color: 'neutral',
    fontSize: 'base',
    fontWeight: 'normal',
    quiet: true,
  },
}

export const FontUtilities: Story = {
  render: () => (
    <div className='flex gap-4'>
      <Anchor
        href='#'
        color='primary'
        fontSize='sm'
        fontWeight='normal'
        quiet={false}
      >
        Small
      </Anchor>
      <Anchor
        href='#'
        color='primary'
        fontSize='base'
        fontWeight='medium'
        quiet={false}
      >
        Medium
      </Anchor>
      <Anchor
        href='#'
        color='primary'
        fontSize='lg'
        fontWeight='bold'
        quiet={false}
      >
        Large
      </Anchor>
      <Anchor
        href='#'
        color='primary'
        fontSize='lg'
        fontWeight='bold'
        quiet={true}
      >
        Large Quiet
      </Anchor>
    </div>
  ),
}
