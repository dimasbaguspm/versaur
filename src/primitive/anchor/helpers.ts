import { cva } from '@/utils/variants'

export const anchorVariants = cva(
  [
    'inline-flex items-center transition-colors focus:outline-none focus-visible:ring-2 focus-visible:ring-offset-2',
    // pseudo selectors for anchor
    'hover:no-underline',
  ],
  {
    variants: {
      color: {
        primary: 'text-primary',
        secondary: 'text-secondary',
        ghost: 'text-ghost',
        danger: 'text-danger',
        neutral: 'text-neutral',
      },
      fontSize: {
        xs: 'text-xs',
        sm: 'text-sm',
        base: 'text-base',
        lg: 'text-lg',
        xl: 'text-xl',
        '2xl': 'text-2xl',
        '3xl': 'text-3xl',
        '4xl': 'text-4xl',
        '5xl': 'text-5xl',
        '6xl': 'text-6xl',
        '7xl': 'text-7xl',
        '8xl': 'text-8xl',
        '9xl': 'text-9xl',
        inherit: '',
        undefined: '',
      },
      fontWeight: {
        thin: 'font-thin',
        extralight: 'font-extralight',
        light: 'font-light',
        normal: 'font-normal',
        medium: 'font-medium',
        semibold: 'font-semibold',
        bold: 'font-bold',
        extrabold: 'font-extrabold',
        black: 'font-black',
        inherit: '',
        undefined: '',
      },
      quiet: {
        true: '',
        false: 'underline underline-offset-2',
      },
    },
    defaultVariants: {
      color: 'primary',
      fontSize: 'base',
      fontWeight: 'medium',
      quiet: false,
    },
  }
)
