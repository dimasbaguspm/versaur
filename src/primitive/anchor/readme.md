# Anchor

Primitive anchor component for Versaur UI library. Renders an accessible `<a>` tag with color roles,
font utilities, and pseudo selector support.

## Usage

```tsx
import { Anchor } from '@/primitive/anchor'

// Default anchor (with underline)
<Anchor
  href='https://versaur.com'
  target='_blank'
  color='primary'
  fontSize='base'
  fontWeight='medium'
>
  Versaur Docs
</Anchor>

// Quiet anchor (no underline)
<Anchor
  href='https://versaur.com'
  target='_blank'
  color='primary'
  fontSize='base'
  fontWeight='medium'
  quiet
>
  Versaur Docs
</Anchor>
```

## Props

- `color`: 'primary' | 'secondary' | 'ghost' | 'danger' | 'neutral' (default: 'primary')
- `fontSize`: Tailwind font size utility (default: 'base')
- `fontWeight`: Tailwind font weight utility (default: 'medium')
- `quiet`: boolean â€” removes underline style when true (default: false)
- All standard anchor (`<a>`) props

## Pseudo Selectors & Variants

- `:hover` and `:visited` states are styled according to color and web standards
- `quiet` disables underline and hover underline

## Accessibility

- Uses native anchor semantics
- Focus ring and hover states for clarity
- Follows WCAG 2.1 AA contrast

## Testing

- Unit tests use `composeStories` from Storybook
- Snapshot and accessibility assertions included

## Storybook

- Stories demonstrate all color roles, font utilities, pseudo selectors, and quiet variant

---

For more details, see the Versaur design system and contribution guidelines.
