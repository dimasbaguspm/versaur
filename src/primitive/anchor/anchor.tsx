import { forwardRef } from 'react'
import type { AnchorProps } from './types'
import { anchorVariants } from './helpers'
import { cn } from '@/utils/cn'

/**
 * Anchor primitive component for rendering links
 * Supports fontSize and fontWeight, and covers pseudo selectors (:hover, :visited)
 * @see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a
 */
export const Anchor = forwardRef<HTMLAnchorElement, AnchorProps>(
  (
    {
      children,
      className,
      color = 'primary',
      fontSize = 'base',
      fontWeight = 'medium',
      quiet = false,
      ...props
    },
    ref
  ) => {
    return (
      <a
        ref={ref}
        className={cn(
          anchorVariants({ color, fontSize, fontWeight, quiet }),
          className
        )}
        {...props}
      >
        {children}
      </a>
    )
  }
)
