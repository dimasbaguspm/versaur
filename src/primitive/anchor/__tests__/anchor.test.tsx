import { render } from '@testing-library/react'
import { composeStories } from '@storybook/react'
import * as stories from '../anchor.stories'

describe('Anchor', () => {
  const {
    Primary,
    PrimaryQuiet,
    Secondary,
    SecondaryQuiet,
    Ghost,
    GhostQuiet,
    Danger,
    DangerQuiet,
    Neutral,
    NeutralQuiet,
    FontUtilities,
  } = composeStories(stories)

  it('renders correctly (snapshot)', () => {
    expect(render(<Primary />).asFragment()).toMatchSnapshot()
    expect(render(<PrimaryQuiet />).asFragment()).toMatchSnapshot()
    expect(render(<Secondary />).asFragment()).toMatchSnapshot()
    expect(render(<SecondaryQuiet />).asFragment()).toMatchSnapshot()
    expect(render(<Ghost />).asFragment()).toMatchSnapshot()
    expect(render(<GhostQuiet />).asFragment()).toMatchSnapshot()
    expect(render(<Danger />).asFragment()).toMatchSnapshot()
    expect(render(<DangerQuiet />).asFragment()).toMatchSnapshot()
    expect(render(<Neutral />).asFragment()).toMatchSnapshot()
    expect(render(<NeutralQuiet />).asFragment()).toMatchSnapshot()
  })

  it('renders all color roles and quiet variants', () => {
    expect(render(<Primary />).getByText('Primary Anchor')).toBeInTheDocument()
    expect(
      render(<PrimaryQuiet />).getByText('Primary Quiet Anchor')
    ).toBeInTheDocument()
    expect(
      render(<Secondary />).getByText('Secondary Anchor')
    ).toBeInTheDocument()
    expect(
      render(<SecondaryQuiet />).getByText('Secondary Quiet Anchor')
    ).toBeInTheDocument()
    expect(render(<Ghost />).getByText('Ghost Anchor')).toBeInTheDocument()
    expect(
      render(<GhostQuiet />).getByText('Ghost Quiet Anchor')
    ).toBeInTheDocument()
    expect(render(<Danger />).getByText('Danger Anchor')).toBeInTheDocument()
    expect(
      render(<DangerQuiet />).getByText('Danger Quiet Anchor')
    ).toBeInTheDocument()
    expect(render(<Neutral />).getByText('Neutral Anchor')).toBeInTheDocument()
    expect(
      render(<NeutralQuiet />).getByText('Neutral Quiet Anchor')
    ).toBeInTheDocument()
  })

  it('renders font utilities', () => {
    const { getByText } = render(<FontUtilities />)
    expect(getByText('Small')).toHaveClass('text-sm')
    expect(getByText('Small')).toHaveClass('font-normal')
    expect(getByText('Medium')).toHaveClass('text-base')
    expect(getByText('Medium')).toHaveClass('font-medium')
    expect(getByText('Large')).toHaveClass('text-lg')
    expect(getByText('Large')).toHaveClass('font-bold')
  })

  it('has accessible role and attributes', () => {
    const { getByRole } = render(<Primary />)
    const anchor = getByRole('link')
    expect(anchor).toHaveAttribute('href', '#')
    expect(anchor).toHaveClass('underline')

    const { getAllByRole: getAllByRoleQuiet } = render(<PrimaryQuiet />)
    const [, anchorQuiet] = getAllByRoleQuiet('link')
    expect(anchorQuiet).toHaveAttribute('href', '#')
    expect(anchorQuiet).not.toHaveClass('underline')
  })
})
