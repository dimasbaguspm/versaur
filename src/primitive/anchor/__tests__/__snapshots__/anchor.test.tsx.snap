// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Anchor > renders correctly (snapshot) 1`] = `
<DocumentFragment>
  <a
    class="inline-flex items-center transition-colors focus:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 hover:no-underline text-primary text-base font-medium underline underline-offset-2"
    href="#"
  >
    Primary Anchor
  </a>
</DocumentFragment>
`;

exports[`Anchor > renders correctly (snapshot) 2`] = `
<DocumentFragment>
  <a
    class="inline-flex items-center transition-colors focus:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 hover:no-underline text-primary text-base font-medium"
    href="#"
  >
    Primary Quiet Anchor
  </a>
</DocumentFragment>
`;

exports[`Anchor > renders correctly (snapshot) 3`] = `
<DocumentFragment>
  <a
    class="inline-flex items-center transition-colors focus:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 hover:no-underline text-secondary text-lg font-bold underline underline-offset-2"
    href="#"
  >
    Secondary Anchor
  </a>
</DocumentFragment>
`;

exports[`Anchor > renders correctly (snapshot) 4`] = `
<DocumentFragment>
  <a
    class="inline-flex items-center transition-colors focus:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 hover:no-underline text-secondary text-lg font-bold"
    href="#"
  >
    Secondary Quiet Anchor
  </a>
</DocumentFragment>
`;

exports[`Anchor > renders correctly (snapshot) 5`] = `
<DocumentFragment>
  <a
    class="inline-flex items-center transition-colors focus:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 hover:no-underline text-ghost text-sm font-normal underline underline-offset-2"
    href="#"
  >
    Ghost Anchor
  </a>
</DocumentFragment>
`;

exports[`Anchor > renders correctly (snapshot) 6`] = `
<DocumentFragment>
  <a
    class="inline-flex items-center transition-colors focus:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 hover:no-underline text-ghost text-sm font-normal"
    href="#"
  >
    Ghost Quiet Anchor
  </a>
</DocumentFragment>
`;

exports[`Anchor > renders correctly (snapshot) 7`] = `
<DocumentFragment>
  <a
    class="inline-flex items-center transition-colors focus:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 hover:no-underline text-danger text-xl font-bold underline underline-offset-2"
    href="#"
  >
    Danger Anchor
  </a>
</DocumentFragment>
`;

exports[`Anchor > renders correctly (snapshot) 8`] = `
<DocumentFragment>
  <a
    class="inline-flex items-center transition-colors focus:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 hover:no-underline text-danger text-xl font-bold"
    href="#"
  >
    Danger Quiet Anchor
  </a>
</DocumentFragment>
`;

exports[`Anchor > renders correctly (snapshot) 9`] = `
<DocumentFragment>
  <a
    class="inline-flex items-center transition-colors focus:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 hover:no-underline text-neutral text-base font-normal underline underline-offset-2"
    href="#"
  >
    Neutral Anchor
  </a>
</DocumentFragment>
`;

exports[`Anchor > renders correctly (snapshot) 10`] = `
<DocumentFragment>
  <a
    class="inline-flex items-center transition-colors focus:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 hover:no-underline text-neutral text-base font-normal"
    href="#"
  >
    Neutral Quiet Anchor
  </a>
</DocumentFragment>
`;
